### PR TITLE
Add Hierarchy class

### PR DESCRIPTION
**Pull request by inglesp**
@ghickman please could you review to check that the code + comments in `hierarchy.py` makes sense, and that you're happy with the tests?

### CODE REVIEW

File: codelists/tests/test_hierarchy.py [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+            starting_node="b", sort_key=lambda label: -ord(label)
+        )
+    ) == [
+        ("b", 0, 0, 0, 1),
+        ("e", 1, 0, 1, 1),
+        ("e", 1, 0, 1, -1),
+        ("d", 1, 1, 0, 1),
+        ("d", 1, 1, 0, -1),
+        ("b", 0, 0, 0, -1),
+    ]
```
Comment (ghickman) [inline codelists/tests/test_hierarchy.py:158]: These magic numbers make it really hard to understand what's going on!

File: codelists/tests/test_hierarchy.py [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+        ("d", 1, 0, 1, 1),
+        ("d", 1, 0, 1, -1),
+        ("e", 1, 1, 0, 1),
+        ("e", 1, 1, 0, -1),
+        ("b", 0, 0, 0, -1),
+    ]
+
+    assert list(
+        hierarchy.walk_depth_first_as_tree(
+            starting_node="b", sort_key=lambda label: -ord(label)
```
Comment (ghickman) [inline codelists/tests/test_hierarchy.py:149]: What does `-ord(label)` do here?

**Review by ghickman — CHANGES_REQUESTED** [review]
Looking good!  This all feels like a much better abstraction now.  The main change is with `NodeWithMetadata` which feels a little too opaque.

File: codelists/hierarchy.py [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+NodeWithMetadata = namedtuple(
+    "NodeWithMetadata",
+    [
+        "label",  # the node's label
+        "depth",  # the length of the shorted path from the root to the node
+        "left_ix",  # the index of the node in the sorted list of its siblings
+        "right_ix",  # the right-index of the node in the sorted list of its siblings
+        "direction",  # 1 when entering a node; -1 when exiting it
+    ],
+)
```
[the remark this answers is not in the document]
Comment (inglesp) [reply codelists/hierarchy.py:261]: What do you think about using kwargs when building a `NodeWithMetadata`?  As in:

```
NodeWithMetadata(label=starting_node, depth=0, left_ix=0, right_ix=0, direction=1)
```

[the remark this answers is not in the document]
Comment (ghickman) [reply codelists/hierarchy.py:261]: Yes please!

File: codelists/tests/test_hierarchy.py [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+        ("c", "f"),
+        ("d", "g"),
+        ("d", "h"),
+        ("e", "h"),
+        ("e", "i"),
+        ("f", "i"),
+        ("f", "j"),
+    ]
+
+    return Hierarchy("a", edges)
```
[the remark this answers is not in the document]
Comment (inglesp) [reply codelists/tests/test_hierarchy.py:53]: Thanks!

File: codelists/hierarchy.py [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+
+        yield NodeWithMetadata(starting_node, 0, 0, 0, 1)
+        yield from helper(starting_node, 1)
+        yield NodeWithMetadata(starting_node, 0, 0, 0, -1)
+
+    def filter_to_ultimate_ancestors(self, nodes):
+        """Return subset of nodes which have no ancestors in given set of nodes.
+        """
+
+        return {node for node in nodes if not self.ancestors(node) & nodes}
```
[the remark this answers is not in the document]
Comment (inglesp) [reply codelists/hierarchy.py:112]: What about

```
"""Given a set of nodes, return subset which have no ancestors in the set.
"""
```

?

File: codelists/tests/test_hierarchy.py [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+    }
+
+    #        a
+    #       / \
+    #      .   .
+    #     / \ / \
+    #    d   e   f
+    assert hierarchy.filter_to_ultimate_ancestors({"a", "b", "c", "d", "e", "f"}) == {
+        "a"
+    }
```
[the remark this answers is not in the document]
Comment (inglesp) [reply codelists/tests/test_hierarchy.py:190]: It'd got different arguments.  Or are you asking why it's worth having a separate test?

File: codelists/hierarchy.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ghickman) [reply codelists/hierarchy.py]: Much clearer, thanks

File: codelists/tests/test_hierarchy.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ghickman) [reply codelists/tests/test_hierarchy.py:198]: Ideally they'd be different tests but I don't feel strongly about it.

However lines 178 and 188 are identical, is that expected?  Based on the comments on lines 183-187 I would have expected line 188 to not have `"c"` and `"d"` in it.  Am I missing something or was that unintended?

[the remark this answers is not in the document]
Comment (inglesp) [reply codelists/tests/test_hierarchy.py:198]: So they are.  I was going by the comments :yikes:

File: codelists/tests/test_hierarchy.py (hunk 2 above)
Comment (inglesp) [reply codelists/tests/test_hierarchy.py:157]: It sorts the labels in reverse alphabetic order.  Does it need a comment?

**Review by ghickman — APPROVED** [review]

File: codelists/tests/test_hierarchy.py (hunk 2 above)
Comment (ghickman) [reply codelists/tests/test_hierarchy.py:157]: Yes, please.  I've read the docs on `ord` at least 3 times trying to get my head around this.  I think I got there in the end but your comment succinctly captured it in a way that I'd not arrived at so probably for the best!

Comment (inglesp) [reply codelists/tests/test_hierarchy.py:158]: Addressed in cb39649fbd4fafcb72910aef717c7339e1944a0f.

**Pull request merged**